lib/control/src/grounder.cc

```diff
@@ -369,7 +372,89 @@ void Grounder::accept(Ground::ProfileNode::Visitor const &visit) const {
     impl_->profile.accept(visit);
 }
 
-auto Grounder::ground(Input::ProgramParamVec const &params, Ground::ScriptCallback *context) -> bool {
+class GroundHandle::Impl {
+  public:
+    Impl() = default;
+
+    Impl(Impl const &) = delete;
+    Impl(Impl &&other) noexcept = delete;
+    auto operator=(Impl const &) -> Impl & = delete;
+    auto operator=(Impl &&other) noexcept -> Impl & = delete;
+
+    void start(Grounder &grd, Input::ProgramParamVec params, Ground::ScriptCallback *context) {
+        assert(!future_.valid());
+        auto task = std::packaged_task<GroundResult()>{
+            [this, &grd, params = std::move(params), context]() { return grd.ground(params, context, &stop_); }};
+        future_ = task.get_future();
+        thread_ = std::thread{std::move(task)};
```

Comment:
AFAICS, you could get rid of the explicit thread and the packaged_task and just use `std::async` with `launch::async`.

The dtor would then basically just need to call `cancel()`.